internal/cri/io/streaming.go

```diff
@@ -0,0 +1,126 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package io
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/containerd/ttrpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	streamingapi "github.com/containerd/containerd/v2/core/streaming"
+	"github.com/containerd/containerd/v2/core/streaming/proxy"
+	"github.com/containerd/containerd/v2/core/transfer/streaming"
+	"github.com/containerd/containerd/v2/pkg/shim"
+)
+
+// ioStream is a stream created by streaming api for io transfer
+// we add a field c as io.Closer because we do connect the streaming server
+// and create a client everytime we create a stream. so we need to close
+// the connection if the stream is closed.
+type ioStream struct {
+	streamingapi.Stream
+	c io.Closer
+}
+
+func (i *ioStream) Close() error {
+	if err := i.Stream.Close(); err != nil {
+		return err
+	}
+	return i.c.Close()
```

Comment:
nit: Should we try to close both and return consolidated error?

```go
errors.Join(i.Stream.Close(), i.c.Close())
```

Reply:
done